test/formulon.spec.js

```diff
@@ -353,4 +353,1407 @@ describe('Formulon', () => {
+                dataType: 'text',
+                value: 'funeral',
+                options: {
+                  length: 7,
+                },
+              },
+              {
+                type: 'literal',
+                dataType: 'text',
+                value: 'fun',
+                options: {
+                  length: 3,
+                },
+              },
+            ],
+          };
+
+          expect(ast('CONTAINS("funeral", "fun")')).to.deep.equal(expected);
+          expect(ast('contains("funeral", "fun")')).to.deep.equal(expected);
+        });
+      });
+    });
+
+    context('Quotes', () => {
+      context('Single Quotes', () => {
+        it('returns expected result', () => {
+          const expected = {
+            type: 'literal',
+            value: 'string',
+            dataType: 'text',
+            options: {
+              length: 6,
+            },
+          };
+
+          expect(ast("'string'")).to.deep.equal(expected);
+        });
+
+        context('Double Quotes in Single Quotes', () => {
+          it('returns expected result', () => {
+            const expected = {
+              type: 'literal',
+              value: '"string"',
+              dataType: 'text',
+              options: {
+                length: 8,
+              },
+            };
+
+            expect(ast("'\"string\"'")).to.deep.equal(expected);
+          });
+        });
+      });
+
+      context('Double Quotes', () => {
+        it('returns expected result', () => {
+          const expected = {
+            type: 'literal',
+            value: 'string',
+            dataType: 'text',
+            options: {
+              length: 6,
+            },
+          };
+
+          expect(ast('"string"')).to.deep.equal(expected);
+        });
+
+        context('Single Quotes in Double Quotes', () => {
+          it('returns expected result', () => {
+            const expected = {
+              type: 'literal',
+              value: "'string'",
+              dataType: 'text',
+              options: {
+                length: 8,
+              },
+            };
+
+            expect(ast("\"'string'\"")).to.deep.equal(expected);
+          });
+        });
+      });
+    });
+
+    context('empty input', () => {
+      const expected = {};
+
+      context('null input', () => {
+        it('returns empty string', () => {
+          expect(ast(null)).to.deep.eq(expected);
+        });
+      });
+
+      context('undefined input', () => {
+        it('returns empty string', () => {
+          expect(ast(undefined)).to.deep.eq(expected);
+        });
+      });
+
+      context('empty string input', () => {
+        it('returns empty string', () => {
+          expect(ast('')).to.deep.eq(expected);
+        });
+      });
+
+      context('whitespace only input', () => {
+        it('returns empty string', () => {
+          expect(ast('   ')).to.deep.eq(expected);
+        });
+      });
+    });
+  });
```

Comment:
Are these the exact same tests as for `parse`. Can we just iterate over the two functions and run the same test?

Reply:
The `empty input` tests are the same as those for `parse` with the exception of the expected return value—it's an empty object for `ast` and `buildLiteralFromJs('')` for `parse`. The rest of the tests are the exact same tests as `build` in `ast.spec.js`, just the method name is different.

Reply:
The `empty input` tests are the same as those for `parse` with the exception of the expected return value—it's an empty object for `ast` and `buildLiteralFromJs('')` for `parse`. The rest of the tests are the exact same tests as `build` in `ast.spec.js`, just the name of the method being tested is different.

Reply:
Yes, that's what I mean.

Is it possible to do something like this (inspired by [Stackoverflow](https://stackoverflow.com/a/48354493/1087469)):

(Careful pseudocode)

```javascript

const Handler= {
  build,
  ast
};


["build", "ast"].forEach(methodName => {
    context('Trimming', () => {
       it('parses AST correctly with trailing whitespace', () => {
         const expected = {
           type: 'literal',
           value: 1,
           dataType: 'number',
           options: {
             length: 1,
             scale: 0,
           },
         };
         expect(Handler[methodName]('1 ')).to.deep.equal(expected);
       });

  ...
})
```

And then additionally add the tests that are unique to each method.

Reply:
Got it. I opted to create a shared method called in each `ast.spec` and `formulon.spec` and pass in the handler. The new method is in `test/shared.js` —let me know if you'd prefer a different name or location.